### PR TITLE
Add SparkPost mailer

### PR DESCRIPTION
**Pull request by simshaun**
One thing I wasn't sure about is what to do with `['options' => ['transactional' => true]]` in the payload data.

It's explained on https://developers.sparkpost.com/api/transmissions.html#transmissions-create-post

In my case, all mail is transactional, but others using Stampie may be using it for marketing purposes.

Is there a clean way of making that parameter configurable? Maybe in the message metadata, but that feels sort of dirty.

### CODE REVIEW

**Comment by simshaun** [comment]
One thing I wasn't sure about is what to do with `['options' => ['transactional' => true]]` in the payload data.

File: lib/Mailer/SparkPost.php [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        ];
+    }
+
+    protected function format(MessageInterface $message)
+    {
+        $from = $this->normalizeIdentity($message->getFrom());
+
+        $parameters = [
+            'options' => [
+                'transactional' => true,
```
Comment (simshaun) [inline lib/Mailer/SparkPost.php]: Is there a clean way of making this configurable?

**Comment by henrikbjorn** [comment]
We have done some `*AwareInterface`s maybe, one should be added for options such as this. Or you can "abuse" the `MetadataAwareInterface`

**Comment by simshaun** [comment]
Something like an `OptionsAwareInterface` would be nice to cover options any mailer may need, as long as there is per-adapter documentation (e.g. `docs/sparkpost.md`) on possible options.

I can go ahead and take care of this if you agree on going that route.

**Comment by simshaun** [comment]
Bump. Do you think an `OptionsAwareInterface` is a good idea?

**Comment by Nyholm** [comment]
I think this PR is fine as it is. We can add the configurable options as a feature in a later PR. 

Is that okey with you?

File: lib/Mailer/SparkPost.php [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+        // 4xx will contain error information in the body encoded as JSON
+        if (!in_array($response->getStatusCode(), range(400, 429), false)) {
+            throw $httpException;
+        }
+
+        throw new ApiException($response->getContent(), $httpException);
+    }
+
+    protected function getAttachmentContent(Attachment $attachment)
```
Comment (stof) [inline lib/Mailer/SparkPost.php]: should be private

File: lib/Mailer/SparkPost.php [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+        return json_encode($parameters);
+    }
+
+    protected function handle(ResponseInterface $response)
+    {
+        $httpException = new HttpException($response->getStatusCode(), $response->getStatusText());
+
+        // 4xx will contain error information in the body encoded as JSON
+        if (!in_array($response->getStatusCode(), range(400, 429), false)) {
```
Comment (stof) [inline lib/Mailer/SparkPost.php]: use `<` and `>=` instead of generating an array with 30 values

File: lib/Mailer/SparkPost.php [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            /*
+             * IMPORTANT:
+             * Do not set address.name!
+             * SparkPost will wrap the entire header_to value, e.g. Bob <bob@example.com, john@example.com>
+             */
+
+            $parameters['recipients'][] = [
+                'address' => [
+                    'email' => $recipient->getEmail(),
+                    'header_to' => $toIdentityString,
```
Comment (stof) [inline lib/Mailer/SparkPost.php]: Reading the source of the official SparkPost Node SDK, this should be a bit different:

- for normal recipients, use `email` and `name` in the JSON recipients
- for CC and BCC, use `email` and `header_to`

File: lib/Mailer/SparkPost.php (hunk 1 above)
Comment (stof) [reply lib/Mailer/SparkPost.php]: I suggest leaving it out for now. And then, we could make it a configuration setting of the mailer itself.

If you want to send both transactional and marketing emails, you would then have to instantiate 2 different mailers

File: lib/Mailer/SparkPost.php (hunk 4 above)
Comment (simshaun) [reply lib/Mailer/SparkPost.php]: It's been a while, so I don't remember the details, but I remember having issues with how SparkPost was building the header. I'll take a look again when I get a chance.

File: lib/Mailer/SparkPost.php (hunk 2 above)
Comment (simshaun) [reply lib/Mailer/SparkPost.php]: I mimicked the Postmark and Mandrill mailers here. Making it private breaks the test at https://github.com/Stampie/Stampie/pull/83/files#diff-7404c8f30d9034dc95f40bd25167575dR183, but I'll try to figure out how to get around it. Might be easiest just to create dummy fixture files. Ideas?

Comment (stof) [reply lib/Mailer/SparkPost.php]: creating a dummy file is better than writing tests against a partial mock (as it means part of the code is untested).
Rewriting other tests is actually on my TODO list

**Comment by Nyholm** [comment]
Could you rebase on master?

File: lib/Mailer/SparkPost.php (hunk 4 above)
Comment (simshaun) [reply lib/Mailer/SparkPost.php]: Looking at why I had it this way again. In the following scenarios, a message has the following "To" recipients: **john@example.com** & **sally@example.com**

# Scenario A:

```php
$parameters['recipients'] = [
  ['address' => ['email' => 'john@example.com', 'name' => 'John']],
  ['address' => ['email' => 'sally@example.com', 'name' => 'Sally']],
];
```

## Scenario A "To" headers

**John's Inbox** - `John <john@example.com>`
**Sally's Inbox** - `Sally <sally@example.com>`

This was undesired behavior for us. We wanted the To header to be `John <john@example.com>,Sally <sally@example.com>`. In other words, show everybody who received the email. If we wanted to truly mimic Scenario A's behavior, then we would loop each recipient, call `->setTo($recipient)` and send.

# Scenario B:

```php
$parameters['recipients'] = [
  ['address' => [
    'email' => 'john@example.com', 
    'header_to' => 'John <john@example.com>,Sally <sally@example.com>'
  ]],
  ['address' => [
    'email' => 'sally@example.com', 
    'header_to' => 'John <john@example.com>,Sally <sally@example.com>'
  ]],
];
```

## Scenario B "To" headers

**John's Inbox** - `John <john@example.com>,Sally <sally@example.com>`
**Sally's Inbox** - `John <john@example.com>,Sally <sally@example.com>`

This is what we wanted. Each recipient receives the email and is able to see everybody else that it was sent to in the To header.

## Why not set the 'name' key in Scenario B?

If `name` is provided, SparkPost mangles the To header like so:

`"John <John" <john@example.com>,"Sally >" <sally@example.com>`

File: lib/Mailer/SparkPost.php (hunk 1 above)
Comment (simshaun) [reply lib/Mailer/SparkPost.php]: I've added an argument to the mailer constructor to control this parameter.

**Comment by Nyholm** [comment]
Awesome. 

To make sure Travis is happy, you need to do something like: 
```php
        if (method_exists($this, 'expectException')) {
            $this->expectException('\InvalidArgumentException');
        } else {
            $this->setExpectedException('\InvalidArgumentException');
        }
```

File: lib/Mailer/SparkPost.php [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+/**
+ * Mailer to be used with SparkPost's API.
+ *
+ * @author Shaun Simmons <shaun@envysphere.com>
+ */
+class SparkPost extends Mailer
+{
+    private $transactional;
+
+    public function __construct(HttpClient $httpClient, $serverToken, $transactional = true)
```
Comment (Nyholm) [inline lib/Mailer/SparkPost.php]: Lets add function docs

**Comment by Nyholm** [comment]
Thank you for your work!

**Pull request merged**